src/triage/predictlist/__init__.py

```diff
@@ -0,0 +1,610 @@
+        )
+        cohort_table_generator.generate_entity_date_table(as_of_dates=[dt_from_str(as_of_date)])
+       
+    def get_collate_aggregations(self, as_of_date, state_table):
+        collate_aggregations = self.feature_generator.aggregations(
+            feature_aggregation_config=self.experiment_config['feature_aggregations'],
+            feature_dates=[as_of_date],
+            state_table=state_table
+        )
+        return collate_aggregations
+
+    def get_feature_dict_and_imputation_task(self, collate_aggregations, model_id):
+        (train_matrix_uuid, matrix_metadata) = train_matrix_info_from_model_id(self.db_engine, model_id)
+        reconstructed_feature_dict = FeatureGroup()
+        imputation_table_tasks = OrderedDict()
+
+        for aggregation in collate_aggregations:
+            feature_group, feature_names = get_feature_names(aggregation, matrix_metadata)
+            reconstructed_feature_dict[feature_group] = feature_names
+
+            # Make sure that the features imputed in training should also be imputed in production
+            
+            features_imputed_in_train = get_feature_needs_imputation_in_train(aggregation, feature_names)
+            
+            features_imputed_in_production = get_feature_needs_imputation_in_production(aggregation, self.db_engine)
+
+            total_impute_cols = set(features_imputed_in_production) | set(features_imputed_in_train)
+            total_nonimpute_cols = set(f for f in set(feature_names) if '_imp' not in f) - total_impute_cols
+            
+            task_generator = self.feature_generator._generate_imp_table_tasks_for
+            
+            imputation_table_tasks.update(task_generator(
+                aggregation,
+                impute_cols=list(total_impute_cols),
+                nonimpute_cols=list(total_nonimpute_cols)
+                )
+            )
+        return reconstructed_feature_dict, imputation_table_tasks
+    
+    def retrain(self, prediction_date):
+        """Retrain a model by going back one split from prediction_date, so the as_of_date for training would be (prediction_date - training_label_timespan)
+        
+        Args:
+            prediction_date(str) 
+        """
+        prediction_date = dt_from_str(prediction_date)
+        as_of_date = datetime.strftime(prediction_date - convert_str_to_relativedelta(self.training_label_timespan), "%Y-%m-%d")
+ 
+        retrain_definition = {
+            'first_as_of_time': dt_from_str(as_of_date),
+            'last_as_of_time': dt_from_str(as_of_date),
+            'matrix_info_end_time': prediction_date,
+            'as_of_times': [dt_from_str(as_of_date)],
+            'training_label_timespan': self.training_label_timespan,
+            'training_as_of_date_frequency': self.experiment_config['temporal_config']['training_as_of_date_frequencies'],
+            'max_training_history': self.experiment_config['temporal_config']['max_training_histories'][0],
+        }
```

Comment:
This doesn't look quite right to me -- maybe it's just a naming issue, but it looks like the first and last `as_of_date`s in the training set are getting set to the same thing here, regardless of the `max_training_history`? Also, shouldn't we be taking the `max_training_history` associated with this particular model group and not just the first one we find in the associated experiment config?

I think a better bet here might be to find reasonable start/end times and run timechop to get the right set of `as_of_dates`

If it's useful, [here's an example](https://github.com/dssg/johnson-county-ddj/blob/master/triage_pipeline/run_trial_predict.py#L97) of how we were doing something similar in JoCo, though with some inferring of the dates from the data, though we also use a single `as_of_date` for both the train and test matrices, so it may need some modification.

Reply:
I created `get_temporal_config_for_retrain` similar to what you did in joco for create a temporal config for timechop. For `max_training_history`, I get it from the sample temporal config.